# CapStone Project

## 1. This section will scrape Toronto data from  Wiki page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [27]:
import pandas as pd
print ('Panda for dataframe, Urlopen to read Url data and BeutifulSoup for web scraping!')


Panda for dataframe, Urlopen to read Url data and BeutifulSoup for web scraping!


In [28]:
from urllib.request import urlopen
from bs4 import BeautifulSoup


url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

html = urlopen(url).read()
soup = BeautifulSoup(html)
soup.prettify()
# print(soup.prettify())

table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df.head(20)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   
15        M5C  Downtown Toronto   
16        M6C              York   
17        M9C         Etobicoke   
18        M1E       Scarborough   
19        M4E      East Toronto   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                    Don Mills South  
14                                   Woodbine Heights  
15                                     St. James Town  
16                                 Humewood-Cedarvale  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                  Guildwood, Morningside, West Hill  
19                                        The Beaches

In [29]:
df.shape

(103, 3)

#### This dataframe has 103 rows and 3 columns

#    # 2. Get the latitude and the longitude coordinates 

In [43]:
latLong =pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv')
latLong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [65]:
latLong.rename(columns={'Postal Code':'PostalCode'},inplace=True)
latLong.rename(columns={'Postalcode':'Postalcode'},inplace=True)
latLong.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [62]:

latLong.shape

(103, 3)

In [66]:
combined =  pd.merge(df, latLong, on='PostalCode')
combined.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

## 3.Explore and cluster the neighborhoods in Toronto.

In [75]:
dfToronto=combined[combined['Borough'].str.contains("Toronto")]
dfToronto.head()

PostalCode           Borough               Neighborhood   Latitude  \
2         M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
9         M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
15        M5C  Downtown Toronto             St. James Town  43.651494   
19        M4E      East Toronto                The Beaches  43.676357   
20        M5E  Downtown Toronto                Berczy Park  43.644771   

    Longitude  
2  -79.360636  
9  -79.378937  
15 -79.375418  
19 -79.293031  
20 -79.373306

In [76]:
CLIENT_ID = 'MF4LH1MWNCZ4SJO3X0UADY5O02ODRKRLF4SGPDJD4RJWIX0B'
CLIENT_SECRET='3UE1L2LSAVVM3YGAGRJSGPLSADRY305NPXTES0UKAQPYHH5I'
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [77]:
toronto_venues = getNearbyVenues(names=dfToronto['Neighborhood'],
                                   latitudes=dfToronto['Latitude'],
                                   longitudes=dfToronto['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

In [78]:
print(toronto_venues.shape)
toronto_venues.head()

(1603, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

In [79]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


In [80]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0            0               0        0                   0             0   
1            0               0        0                   0             0   
2            0               0        0                   0             0   
3            0               0        0                   0             0   
4            0               0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Tibetan Restaurant  Toy / Game Store  \
0             0  ...                 0                   0                 0   
1             0  ...                 0                   0                 0   
2             0  ...                 0                   0                 0   
3             0  ...                 0                   0                 0   
4             0  ...                 0                   0                 0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 235 columns]

examine the new dataframe size.


In [81]:
toronto_onehot.shape

(1603, 235)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [82]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
3                                  Central Bay Street     0.014493   
4                                            Christie     0.000000   
5                                Church and Wellesley     0.025641   
6                      Commerce Court, Victoria Hotel     0.010000   
7                                          Davisville     0.000000   
8                                    Davisville North     0.000000   
9                        Dufferin, Dovercourt Village     0.000000   
10                                     Enclave of M4L     0.066667   
11                                     Enclave of M5E     0.010309   
12             First Canadian Place, Underground city     0.000000   
13                           Forest Hill North & West     0.000000   
14                           Garden District, Ryerson     0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000   
16                      High Park, The Junction South     0.000000   
17                     India Bazaar, The Beaches West     0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000   
19                                      Lawrence Park     0.000000   
20                           Little Portugal, Trinity     0.023810   
21                        Moore Park, Summerhill East     0.000000   
22                                 North Toronto West     0.050000   
23                             Parkdale, Roncesvalles     0.000000   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.028571   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                    Studio District     0.027778   
32  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
33                The Annex, North Midtown, Yorkville     0.000000   
34                                        The Beaches     0.000000   
35                                 The Danforth  East     0.000000   
36                       The Danforth West, Riverdale     0.023810   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.027778   

    Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0         0.000000   0.0000              0.0000        0.0000          0.0000   
1         0.000000   0.0000              0.0000        0.0000          0.0000   
2         0.000000   0.0625              0.0625        0.0625          0.0625   
3         0.000000   0.0000              0.0000        0.0000          0.0000   
4         0.000000   0.0000              0.0000        0.0000          0.0000   
5         0.012821   0.0000              0.0000        0.0000          0.0000   
6         0.000000   0.0000              0.0000        0.0000          0.0000   
7         0.000000   0.0000              0.0000        0.0000          0.0000   
8         0.000000   0.0000              0.0000        0.0000          0.0000   
9         0.000000   0.0000              0.0000        0.0000          0.0000   
10        0.000000   0.0000              0.0000        0.0000          0.0000   
11        0.000000   0.0000              0.0000        0.0000          0.0000   
12        0.000000   0.0000              0.0000        0.0000          0.0000   
13        0.000000   0.0000              0.0000        0.0000          0.0000   
14        0.000000   0.00

In [83]:
toronto_grouped.shape

(39, 235)

Let's print each neighborhood along with the top 5 most common venues

In [84]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0  Cocktail Bar  0.05
1        Bakery  0.05
2   Coffee Shop  0.05
3   Cheese Shop  0.04
4    Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2          Bakery  0.08
3     Coffee Shop  0.08
4      Restaurant  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.19
1    Airport Terminal  0.12
2               Plane  0.06
3             Airport  0.06
4  Airport Food Court  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.16
1      Sandwich Place  0.06
2                Café  0.06
3          Restaurant  0.04
4  Italian Restaurant  0.04


----Christie----
                venue  freq
0       Grocery Store  0.25
1                Café  0.19
2                Park  0.12
3           Nightclub  0.06
4  Itali

In [85]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [86]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                Bakery   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Coffee Shop    Seafood Restaurant   
1           Coffee Shop                Bakery        Breakfast Spot   
2      Airport Terminal         Boat or Ferry   Rental Car Location   
3        Sandwich Place                  Café            Restaurant   
4                  Café                  Park           Candy Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant        Farmers Market              Beer Bar   
1               Stadium         Burrito Place            Restaurant   
2                   Bar                 Plane           Coffee Shop   
3    Italian Restaurant   Japanese Restaurant           Salad Place   
4           Coffee Shop            Restaurant    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Pharmacy           Cheese Shop                  Diner  
1          Climbing Gym             Pet Store  Performing Arts Venue  
2       Harbor / Marina      Sculpture Garden           Airport Gate  
3       Bubble Tea Shop          Burger Joint                    Spa  
4    Italian Restaurant             Nightclub             Baby Store

## 4. Cluster Neighborhoods

In [89]:
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

In [90]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood   Latitude  \
2         M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
9         M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
15        M5C  Downtown Toronto             St. James Town  43.651494   
19        M4E      East Toronto                The Beaches  43.676357   
20        M5E  Downtown Toronto                Berczy Park  43.644771   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  -79.360636               2           Coffee Shop                Bakery   
9  -79.378937               2           Coffee Shop        Clothing Store   
15 -79.375418               2           Coffee Shop                  Café   
19 -79.293031               0     Health Food Store                 Trail   
20 -79.373306               2                Bakery          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                    Pub                  Park        Breakfast Spot   
9                  Hotel   Japanese Restaurant       Bubble Tea Shop   
15            Restaurant        Clothing Store                Bakery   
19                   Pub         Women's Store               Dog Run   
20           Coffee Shop    Seafood Restaurant            Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2                   Café               Theater                   Spa   
9         Cosmetics Shop                  Café  Fast Food Restaurant   
15          Cocktail Bar        Cosmetics Shop              Creperie   
19          Dessert Shop                 Diner        Discount Store   
20        Farmers Market              Beer Bar              Pharmacy   

   9th Most Common Venue 10th Most Common Venue  
2             Beer Store             Shoe Store  
9       Department Store       Ramen Restaurant  
15        Lingerie Store    Moroccan Restaurant  
19   Distribution Center             Donut Shop  
20           Cheese Shop                  Diner

In [93]:
!pip install folium
import folium # map rendering library

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 94 kB 4.6 MB/s  eta 0:00:01


In [97]:
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import numpy as np # library to handle data in a vectorized manner
import requests # library to handle requests
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.7-main

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |           1_llvm           5 KB  conda-forge
    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge
    _pytorch_select-0.2        |            gpu_0           2 KB
    absl-py-0.13.0             |     pyhd8ed1ab_0          97 KB  conda-forge
    aiohttp-3.7.4.post0        |   py37h5e8e339_0         

cached-property-1.5. | 4 KB      | ##################################### | 100% 
ipython-7.25.0       | 1.1 MB    | ##################################### | 100% 
prometheus_client-0. | 46 KB     | ##################################### | 100% 
importlib-metadata-4 | 31 KB     | ##################################### | 100% 
html5lib-1.1         | 89 KB     | ##################################### | 100% 
aiohttp-3.7.4.post0  | 625 KB    | ##################################### | 100% 
pcre-8.45            | 253 KB    | ##################################### | 100% 
libgcc-ng-11.1.0     | 6.7 MB    | ##################################### | 100% 
brunsli-0.1          | 200 KB    | ##################################### | 100% 
libxcb-1.13          | 395 KB    | ##################################### | 100% 
xlrd-2.0.1           | 92 KB     | ##################################### | 100% 
requests-oauthlib-1. | 21 KB     | ##################################### | 100% 
yarl-1.6.3           | 142 K

pillow-8.3.1         | 692 KB    | ##################################### | 100% 
_libgcc_mutex-0.1    | 3 KB      | ##################################### | 100% 
wcwidth-0.2.5        | 33 KB     | ##################################### | 100% 
seaborn-base-0.11.1  | 217 KB    | ##################################### | 100% 
jupyterlab_widgets-1 | 130 KB    | ##################################### | 100% 
charls-2.2.0         | 138 KB    | ##################################### | 100% 
libogg-1.3.4         | 206 KB    | ##################################### | 100% 
libxml2-2.9.12       | 772 KB    | ##################################### | 100% 
pyqt-5.12.3          | 21 KB     | ##################################### | 100% 
nltk-3.6.2           | 1.1 MB    | ##################################### | 100% 
tenacity-8.0.1       | 21 KB     | ##################################### | 100% 
tabulate-0.8.9       | 26 KB     | ##################################### | 100% 
typing_extensions-3. | 28 KB

tk-8.6.10            | 3.2 MB    | ##################################### | 100% 
libpng-1.6.37        | 306 KB    | ##################################### | 100% 
jinja2-3.0.1         | 99 KB     | ##################################### | 100% 
libclang-11.1.0      | 19.2 MB   | ##################################### | 100% 
_py-xgboost-mutex-2. | 8 KB      | ##################################### | 100% 
tensorflow-estimator | 645 KB    | ##################################### | 100% 
freetds-1.1.15       | 2.5 MB    | ##################################### | 100% 
libgfortran4-7.5.0   | 1.3 MB    | ##################################### | 100% 
charset-normalizer-2 | 32 KB     | ##################################### | 100% 
markupsafe-2.0.1     | 22 KB     | ##################################### | 100% 
backports-1.0        | 4 KB      | ##################################### | 100% 
libffi-3.3           | 51 KB     | ##################################### | 100% 
importlib_metadata-4 | 3 KB 

CondaVerificationError: The package for qt located at /opt/conda/pkgs/qt-5.12.9-hda022c4_4
appears to be corrupted. The path 'qml/QtQuick/Controls.2/Imagine/RadioDelegate.qml'
specified in the package manifest cannot be found.

CondaVerificationError: The package for qt located at /opt/conda/pkgs/qt-5.12.9-hda022c4_4
appears to be corrupted. The path 'qml/QtQuick/Controls.2/Imagine/RangeSlider.qml'
specified in the package manifest cannot be found.

CondaVerificationError: The package for qt located at /opt/conda/pkgs/qt-5.12.9-hda022c4_4
appears to be corrupted. The path 'qml/QtQuick/Controls.2/Imagine/RoundButton.qml'
specified in the package manifest cannot be found.

CondaVerificationError: The package for qt located at /opt/conda/pkgs/qt-5.12.9-hda022c4_4
appears to be corrupted. The path 'qml/QtQuick/Controls.2/Imagine/ScrollBar.qml'
specified in the package manifest cannot be found.

CondaVerificationError: The package for qt located at /opt/conda/pkgs/qt-5.12.9-hda022c4_4
appea

ClobberError: The package 'conda-forge/linux-64::numpy-1.20.3-py37h038b26d_0' cannot be installed due to a
path collision for 'lib/python3.7/site-packages/numpy/distutils/setup.py'.
This path already exists in the target prefix, and it won't be removed by
an uninstall action in this transaction. The path appears to be coming from
the package 'defaults/linux-64::numpy-base-1.18.5-py37hde5b4d6_0', which is already installed in the prefix.


ClobberError: The package 'conda-forge/linux-64::numpy-1.20.3-py37h038b26d_0' cannot be installed due to a
path collision for 'lib/python3.7/site-packages/numpy/distutils/site.cfg'.
This path already exists in the target prefix, and it won't be removed by
an uninstall action in this transaction. The path appears to be coming from
the package 'defaults/linux-64::numpy-base-1.18.5-py37hde5b4d6_0', which is already installed in the prefix.


ClobberError: The package 'conda-forge/linux-64::numpy-1.20.3-py37h038b26d_0' cannot be installed due to a
path co

NameError: name 'latitude' is not defined

In [99]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [100]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters